python/lib/client/dmod/client/request_clients.py

```diff
@@ -1,600 +1,823 @@
+            return MaaSDatasetManagementResponse(success=False, reason=reason, message=str(e))
+
         with dest.open('w') as file:
-            for page in range(1, (self.last_response.total_pages + 1)):
-                request = DatasetManagementMessage(action=ManagementAction.DOWNLOAD_DATA, dataset_name=dataset_name,
-                                                   data_location=item_name, page=page)
-                self.last_response: DatasetManagementResponse = await self.async_make_request(request)
-                file.write(self.last_response.file_data)
+            # Do initial request outside of generator
+            await self._transport_client.async_send(data=str(request))
+            async for received_data_msg in self._transfer_receiver():
+                file.write(received_data_msg.data)
+            final_data = await self._transport_client.async_recv()
 
-    async def list_datasets(self, category: Optional[DataCategory] = None) -> List[str]:
-        action = ManagementAction.LIST_ALL if category is None else ManagementAction.SEARCH
-        request = DatasetManagementMessage(action=action, category=category)
-        self.last_response = await self.async_make_request(request)
-        return self._parse_list_of_dataset_names_from_response(self.last_response)
+        try:
+            final_response_json = json.loads(final_data)
+        except Exception as e:
+            msg = f"{self.__class__.__name__} failed with {e.__class__.__name__} parsing `{final_data}` to JSON)"
+            return final_response_type(success=False, reason=f"JSON Parse Failure On Final Response", message=msg)
+
+        final_response = final_response_type.factory_init_from_deserialized_json(final_response_json)
+        if final_response is None:
+            return final_response_type(success=False, reason="Failed to Deserialize Final Response")
+        else:
+            return final_response
+
+    async def upload_dataset_item(self, dataset_name: str, item_name: str, source: Path) -> DatasetManagementResponse:
+        if not source.is_file():
+            return DatasetManagementResponse(success=False, reason="Dataset Upload File Not Found",
+                                             message=f"File {source!s} does not exist")
+        try:
+            request, final_response_type = self._request_prep(dataset_name=dataset_name, item_name=item_name,
+                                                              action=ManagementAction.ADD_DATA)
+        # TODO: (later) implement and use DmodAuthenticationFailure
+        except DmodRuntimeError as e:
+            reason = f'{self.__class__.__name__} Upload Request Auth Failure'
+            return MaaSDatasetManagementResponse(success=False, reason=reason, message=str(e))
 
-    async def upload_to_dataset(self, dataset_name: str, paths: List[Path]) -> bool:
+        chunk_size = 1024
+
+        with source.open('r') as file:
```

Comment:
I think we need to open this in reading bytes, right? Since we dont know what kind of item is being uploaded.

Reply:
Hmm ... yeah, I believe you are correct, and as such we should do similarly for writes as you suggested above.   But this actually becomes a larger undertaking.   All the links in the chain will need to be modified to accommodate, from _DataTransmitMessage.data_ to the declaration and implementations of _DatasetManager.add_data()` (and associated helper methods).  

It probably makes sense to do this as part of #432.  What do you think?

Reply:
Ping again here also.  This probably relates to the earlier item and my old follow-up questions.

Reply:
Yeah I agree. This is a larger undertaking and not directly related so should be in its own PR. 

I opened #486 to track changing `DataTransmitMessage`'s `data` field to a more appropriate type. Im not sure if there are other models off the top of my head that will also need to be changed.

Reply:
It didn't occur to me back in September, but (unless I'm missing something) we can't actually do this because it will break JSON serialization.

Reply:
Yeahhhhh, you are right. I forgot about that. Does that mean you can only upload and download unicode encoded data at this moment? I don't recall seeing any logic to base64 / base85 encode binary data.

Reply:
Just as a sanity check, if we decode and then encode binary data using the "wrong" encoding, but it's the same encoding on both ends, won't we get the same data we started with?

Assuming more discussion and work is needed, I suggest that we defer this problem.  A solution will probably be complex, and we can find workarounds as needed to get datasets into the system (worst case: temporarily opening up MinIO externally and uploading directly).

Reply:
> Just as a sanity check, if we decode and then encode binary data using the "wrong" encoding, but it's the same encoding on both ends, won't we get the same data we started with?

Yeah! You are right. I may be blowing this out of proportion and it might not be a problem after all. I thought to transmit _any_ data to and from the request service (and therefore the data service) it must be in json and therefore in some unicode flavor (e.g. utf-8). So, for example if I wanted to upload a netCDF forcing file to a dataset, _I think_ I would first need to encode and chunk the data using base64 or base85, package that up in json, and send it along, right? I agree that we should move this discussion elsewhere and just focus on getting the PR through lol. Sorry to trying to derail us 😅!